Custom expression builder

JsonHrdwick opened the pull request:
Branch feature for users to be able to create their own custom-built expressions utilizing UserFunctions and mathematical operators

Initialized branch with some basic class structures for implementing this logic. ShuntingYard, ExpressionTrees, and a few operator classes exist but need further refinement

JsonHrdwick commented:
Update:
- Class structures exist for building expression trees using Operations as roots and Constant/Userfunction Nodes as leaves
- Single Arg operations have been implemented with 2 arg conditions as well.
(ex. **Log** with null right child will do _log_10(a)_, if right child is passed in then the method calculates _log_b(a)_ where '_a_' and '_b_' are left and right child respectively)
If this is unnecessary, then I can remove it.
- Built a test isotopic userfunction with a constant value of "10.0" which will be removed later. Adjusted logic in MassSpec class to allow for a 0 StdErr.
- Test function written to ensure the Expression tree populates correctly and evaluates in the proper order. Asserts against known values on the the equation.

Next steps:
- [ ] Implement the drag and drop into the text box for frontend functionality
- [ ] Typing out custom expressions
- [ ] Storing custom expressions in the AnalysisMethod
- [ ] populating custom expressions data in massspec table

bowring commented:
looking good 
add support for parentheses

JsonHrdwick commented:
For some reason, it wouldn't let me add the oracle to git so I had to manually add it via github...

- Tree now evaluates off raw data and returns the matrix of the expression applied to each cycle. 
- Test was adjusted to compare the expression against the entire matrix of data for known correct values.
- Fixed the node assignments for TextFlow
- Fixed indexing issue for TextFlow (Still needs more testing)

Still have plenty to do :)

bowring commented:
check syntax of expression in line 178 of AnalysisMethod.java - seems there is missing operator before first parenthesis.

JsonHrdwick commented:
> check syntax of expression in line 178 of AnalysisMethod.java - seems there is missing operator before first parenthesis.

This is intentional. The omission tests the presumption of a multiplication when lacking an operation before the left parenthesis (ex. "2(x+3)" ) I can remove this if you want but it works as intended in my testing.

The latter issue I will address 😄

JsonHrdwick commented:
Implemented the number cell in the operations listview as suggested. There is a context menu for setting the value such as in squid. I think making the context menu immediately pop up on drop would be a good indicator for how to enter the value. Let me know what you think.
Fixed the issue with data reloading on resaving an existing expression. Had to clear out the MapBlockIdToBlockCyclesRecord from the userfunction to ensure it stored the new data before plotting.
I could not replicate your issue with multiple expressions populating under the same name. If you are on the current commit and still getting this issue, perhaps try resetting your persistantstate file. Let me know if this still needs addressed.

JsonHrdwick commented:
@bowring

I see the issue youre reporting. If you have a solution already, we can try it out after our next meeting. I looked at it a bit and wasn't sure what the best way to approach it was since it had to deal with the way sessions are saved. I would have to look at that code a bit more. 

I did fix the feature for ratio flag in the userfunction (and removing it when the expression is no longer a ratio). 

The report builder now populates expressions properly but I haven't looked at the "faulty data" bug yet for that feature.

JsonHrdwick commented:
@bowring 

Session saved expressions are now handled. User loads a session with mismatched custom expressions in it's analysis and a prompt warns the user and allows them to confirm if they want the session expressions or the method expressions. Implemented a flag in the session class that determines whether the application will save the expressions to the method persistence or not. If the user selects to not refresh the expressions then the saved feature will not save to persistence which ensures they are only local to the session file. If the user refreshes or is not prompted at all, the save will alter the persistence as intended.

Also double checked the ratio flag for single-node expressions and could not replicate the issue of it not being correctly flagged as a ratio. Let me know specifically where you are encountering this issue.

Fixed the report builder CSV generation not adjusting for hidden columns/categories. Reordered the labels to correctly match the data presented (was out of order). If the data is still wrong in the report, I'm not sure what is the cause as I am copying the code from the plots generation. The most I see is a rounding error.

Still would like to move the reports to persistence before publishing

bowring commented:
@JsonHrdwick - please remind me re: "Still would like to move the reports to persistence before publishing"

bowring commented:
do it

also, ratios are not showing correct data in reports - I reset persistent state and started fresh with GHR1 230921 F08b-216

see attached
![ratio](https://github.com/user-attachments/assets/bd651302-0aed-45f3-ac5e-544402dc0d4d)

bowring commented:
looks good

make label wider for method name in header of report builder (minor, see GHR1 230921F03 U-239 )

bowring commented:
the cvs reports should mimic the CLIPBOARD EXPORTS

bowring commented:
bug:  edit expression that has ratio alias and you get an error on save due to alias
bug: can't edit expression name
bug: text mode while editing leaves save button inactive

JsonHrdwick commented:
> bug: edit expression that has ratio alias and you get an error on save due to alias bug: can't edit expression name bug: text mode while editing leaves save button inactive

Fixed Bug one.
Bug 2 is intentional (Cannot edit name of existing expression) Can change this if you want
Bug 3, save is updated when switching back to d&d to prevent the state manager from saving every keystroke as a previous state. Again, can look into a workaround if you want
